apps/jupyter/extensions/notebook-runner/src/utils.ts

```diff
@@ -0,0 +1,147 @@
+  if (isPatched(kernel)) {
+    return;
+  }
+
+  // @ts-expect-error This is a custom property
+  kernel[isPreparedString] = false;
+
+  let resolve: () => void;
+  let reject: () => void;
+
+  // @ts-expect-error This is a custom property
+  kernel[preparedString] = new Promise<void>((res, rej) => {
+    resolve = res;
+    reject = rej;
+  });
+
+  // @ts-expect-error This is a custom property
+  kernel[preparedResolveString] = resolve;
+  // @ts-expect-error This is a custom property
+  kernel[preparedRejectString] = reject;
+};
+
+export const isKernelPrepared = async (
+  kernel: IKernelConnection,
+): Promise<boolean> => {
+  ensureKernelIsPatched(kernel);
+
+  // @ts-expect-error This is a custom property
+  return kernel[isPreparedString];
+};
+
+export const waitForKernelToBePrepared = async (
+  kernel: IKernelConnection,
+): Promise<void> => {
+  ensureKernelIsPatched(kernel);
+
+  // @ts-expect-error This is a custom property
+  await kernel[preparedString];
+};
+
+export const setKernelIsPrepared = (
+  kernel: IKernelConnection,
+  isPrepared: boolean = true,
+): void => {
+  ensureKernelIsPatched(kernel);
+
+  // @ts-expect-error This is a custom property
+  kernel[isPreparedString] = isPrepared;
+
+  if (isPrepared) {
+    // @ts-expect-error This is a custom property
+    const resolve = kernel[preparedResolveString] as () => void;
+    resolve();
+  } else {
+    // @ts-expect-error This is a custom property
+    const reject = kernel[preparedRejectString] as () => void;
+    reject();
+  }
+};
+
+export const addKernelListeners = async (
+  sessionContext: ISessionContext,
+  addListeners: (kernel: IKernelConnection) => Promise<void>,
+): Promise<void> => {
+  console.debug("Waiting for session context to be ready...");
+  await sessionContext.ready;
+
+  while (!sessionContext.session?.kernel) {
+    console.debug("Starting kernel...");
+    await sessionContext.startKernel();
+  }
+
+  const kernel = sessionContext.session?.kernel;
+  await kernel.info;
+  await addListeners(kernel);
+
+  sessionContext.kernelChanged.connect(async (sessionCtx) => {
+    const kernel = sessionCtx.session?.kernel;
+    if (!kernel) {
+      console.warn("Kernel is not available in session context");
+
+      // simply return, the function will be called again when the kernel is changed
+      return;
+    }
+
+    console.log("Kernel changed:", kernel.name);
+    await kernel.info;
+
+    return addListeners(kernel);
+  });
+};
+
+export const writeJsonToVirtualFilesystem = async (
+  kernel: IKernelConnection,
+  path: string,
+  json: unknown,
+): Promise<void> => {
+  const jsonString = JSON.stringify(json);
+  const utf8Bytes = new TextEncoder().encode(jsonString);
+  const binaryString = Array.from(utf8Bytes, (byte) =>
+    String.fromCharCode(byte),
+  ).join("");
+
+  await kernel.requestExecute({
+    // it seems as if the encoding here does not allow us to put JSON.stringify directly
```

Comment:
This is quite weird, have you investigated this in-depth?
Both Python and HTML5 default to UTF-8, which makes encoding issues unlikely.

Reply:
You might want to clarify the suspected issue here, so that future me doesn't make the same stupid assumptions as past me :-)